File: tests/test_windowing.py
from typing import List

import numpy as np
import musher


def normalize(window: List[float]) -> List[float]:
    return [2.0*i/sum(window) for i in window]


def triangular(window: List[float]) -> List[float]:
    size = len(window)
    return [2.0/size*(size/2.0-abs(float(i-(size-1.0)/2.0))) for i in range(size)]


def test_windowing_square():
    inp = [1.] * 1024
    actual_window = musher.windowing(
        inp, musher.square, zero_padding_size=0, zero_phase=False)

    expected_window = normalize(inp)
    assert np.array_equal(actual_window, expected_window)


def test_windowing_custom_triangular():
    """Example of using a custom windowing function (triangular window).

    Custom functions must accept a list of floats and return a list of floats.
    """
    inp = [1.] * 1024
    actual_window = musher.windowing(
        inp, triangular, zero_padding_size=0, zero_phase=False)

    expected_window = normalize(triangular(inp))
    assert np.array_equal(actual_window, expected_window)

File: src/core/test/test_windowing.cpp
#include "gtest/gtest.h"
#include "src/core/test/gtest_extras.h"
#include "src/core/windowing.h"


using namespace musher::core;

/**
 * @brief Square windowing.
 * 
 */
TEST(Windowing, SquareWindowing) {
  std::vector<double> input(1024, 1.);
  unsigned int zero_padding_size = 0;
  bool zero_phase = false;

  std::vector<double> actual_window = Windowing(input, Square, zero_padding_size, zero_phase);

  std::vector<double> expected_window(Normalize(input));

  EXPECT_VEC_EQ(actual_window, expected_window)
}
